backend/src/authentication/handlers.rs

```diff
@@ -0,0 +1,261 @@
+};
+
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct Credentials {
+    username: String,
+    password: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct LoginResponse {
+    user_id: u32,
+    username: String,
+}
+
+impl From<&User> for LoginResponse {
+    fn from(user: &User) -> Self {
+        Self {
+            user_id: user.id(),
+            username: user.username().to_string(),
+        }
+    }
+}
+
+/// Set default session cookie properties
+fn set_default_cookie_properties(cookie: &mut Cookie) {
+    cookie.set_path("/");
+    cookie.set_http_only(true);
+    cookie.set_secure(SECURE_COOKIES);
+    cookie.set_same_site(SameSite::Strict);
+}
+
+/// Login endpoint, authenticates a user and creates a new session + session cookie
+#[utoipa::path(
+    post,
+    path = "/api/user/login",
+    request_body = Credentials,
+    responses(
+        (status = 200, description = "The logged in user id and user name", body = LoginResponse),
+        (status = 401, description = "Invalid credentials", body = ErrorResponse),
+        (status = 500, description = "Internal server error", body = ErrorResponse),
+    ),
+)]
+pub async fn login(
+    State(users): State<Users>,
+    State(sessions): State<Sessions>,
+    jar: CookieJar,
+    Json(credentials): Json<Credentials>,
+) -> Result<impl IntoResponse, APIError> {
+    let Credentials { username, password } = credentials;
+
+    // Check the username + password combination
+    let user = users.authenticate(&username, &password).await?;
+
+    // Remove old sessions, we do this after a login to prevent the necessity of periodical cleanup jobs
+    sessions.delete_old_sessions().await?;
+
+    // Create a new session and cookie
+    let session = sessions.create(user.id(), SESSION_LIFE_TIME).await?;
+
+    // Add the session cookie to the response
+    let mut cookie = session.get_cookie();
+    set_default_cookie_properties(&mut cookie);
+    let updated_jar = jar.add(cookie);
+
+    Ok((updated_jar, Json(LoginResponse::from(&user))))
+}
+
+/// Logout endpoint, deletes the session cookie
+#[utoipa::path(
+    post,
+    path = "/api/user/logout",
+    responses(
+        (status = 200, description = "Successful logout, or user was already logged out"),
+        (status = 500, description = "Internal server error", body = ErrorResponse),
+    ),
+)]
+pub async fn logout(
+    State(sessions): State<Sessions>,
+    jar: CookieJar,
+) -> Result<impl IntoResponse, APIError> {
+    let Some(mut cookie) = jar.get(SESSION_COOKIE_NAME).cloned() else {
+        // no cookie found, return OK
+        return Ok((jar, StatusCode::OK));
+    };
+
+    // Remove session form the database
+    let Some(session) = sessions.get_by_key(cookie.value()).await? else {
+        // no session found, return OK
+        return Ok((jar, StatusCode::OK));
+    };
+
+    sessions.delete(session.session_key()).await?;
```

Comment:
I think we can delete the session from the database without checking if it exists first